server/src/main/java/org/elasticsearch/gateway/LucenePersistedStateFactory.java

```diff
@@ -0,0 +1,660 @@
+        public ClusterState getLastAcceptedState() {
+            return lastAcceptedState;
+        }
+
+        void persistInitialState() throws IOException {
+            // Write the whole state out to be sure it's fresh and using the latest format. Called during initialisation, so that
+            // (1) throwing an IOException is enough to halt the node, and
+            // (2) the index is currently empty since it was opened with IndexWriterConfig.OpenMode.CREATE
+
+            // In the common case it's actually sufficient to commit() the existing state and not do any indexing. For instance, this is
+            // true if there's only one data path on this master node, and the commit we just loaded was already written out by this
+            // version of Elasticsearch. TODO TBD should we avoid indexing when possible?
+            addMetaData(lastAcceptedState);
+            commit(currentTerm, lastAcceptedState.getVersion());
+        }
+
+        @Override
+        public void setCurrentTerm(long currentTerm) {
+            commit(currentTerm, lastAcceptedState.version());
+            this.currentTerm = currentTerm;
+        }
+
+        @Override
+        public void setLastAcceptedState(ClusterState clusterState) {
+            try {
+                if (clusterState.term() != lastAcceptedState.term()) {
+                    assert clusterState.term() > lastAcceptedState.term() : clusterState.term() + " vs " + lastAcceptedState.term();
+                    // In a new currentTerm, we cannot compare the persisted metadata's lastAcceptedVersion to those in the new state, so
+                    // it's simplest to write everything again.
+                    overwriteMetaData(clusterState);
+                } else {
+                    // Within the same currentTerm, we _can_ use metadata versions to skip unnecessary writing.
+                    updateMetaData(clusterState);
+                }
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+
+            commit(currentTerm, clusterState.version());
+            lastAcceptedState = clusterState;
+        }
+
+        /**
+         * Update the persisted metadata to match the given cluster state by removing any stale or unnecessary documents and adding any
+         * updated documents.
+         */
+        private void updateMetaData(ClusterState clusterState) throws IOException {
+            assert lastAcceptedState.term() == clusterState.term();
+            logger.trace("currentTerm [{}] matches previous currentTerm, writing changes only", clusterState.term());
+
+            try (ReleasableDocument globalMetaDataDocument = makeGlobalMetaDataDocument(clusterState)) {
+                for (MetaDataIndexWriter metaDataIndexWriter : metaDataIndexWriters) {
+                    metaDataIndexWriter.deleteGlobalMetaData();
+                    metaDataIndexWriter.addGlobalMetaData(globalMetaDataDocument.getDocument());
```

Comment:
Is it worth separating delete and add? Perhaps just have a single "updateGlobalMetaData" method and use updateDocument, simplifies the code, and has no performance impact

Reply:
Sure, see 7fa29f96ad9c2d92cfff9f38a53026f0af5917d5.